### DISCUSSION TITLE
Welcome to xmb-linux Discussions!

### DISCUSSION THREAD

**Discussion by ProximalElk6186** [ProximalElk6186/xmb-linux#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by 5d-code** [comment]
hey i am working on adding a clock thats rendered inside of the canvas instead of HTML, a declarative method to make menus, adding callbacks when an item is clicked/pressed, gamepad support and adding support for multiple scenes. when we are done with the p5js implementation we can migrate to vanilla js by just using canvas or even migrate to SDL and just use node? this would allow us to make the wave effect with some OpenGL (we can do it also with WebGL) and make text glow when its selected

**Reply by ProximalElk6186** [reply]
Great job!
About clock: Yeah, it definately needs to be rendered inside of the canvas.
About migration: let's schedule the decision. As for SDL and just using Node, well, AFAIK, it requires some C/C++ programming knowledge. I am familiar with C-like languages, so it shouldn't be a huge problem. To sum up: Let's do SDL and just use Node, but later; we should definately first make it work in p5js.

**Reply by ProximalElk6186** [reply]
Also, maybe make a fork of this project, do all the changes and then do a pull request?

**Reply by 5d-code** [reply]
ok I'll make a fork, also, we will make it in p5js because I don't use a lot of functions from it and I can just make a simple SDL wrapper for node and name the functions after the p5js ones so its drop in replacement

**Reply by ProximalElk6186** [reply]
Ok

**Comment by 5d-code** [comment]
ok quick question: will we make a recreation of PS3's XMB down to the last string, or will we make a recreation of the XMB UI and then adapt it for computer usage? like for example we can make this like a media center application like kodi?????

**Reply by ProximalElk6186** [reply]
i think we will adapt it for computer usage.

**Comment by 5d-code** [comment]
its ready I migrated to typescript and there is a new api
you can now add callbacks to buttons and custom icons for each one
now all that's left is descriptions & submenus (oh no) from the functionality side
I don't think glowing text will be possible on the web without the use of webgl,
since glowing text is achieved by putting a gaussian blurred version of the text under the normal text
and since the ps3 xmb pulsates the glowing intensity aka rerendering the blur each frame
we could add memoization but still it would be very laggy

**Reply by 5d-code** [reply]
![image](https://github.com/user-attachments/assets/ab554ca4-532c-4370-99f3-266018fa68dc)

**Reply by 5d-code** [reply]
PS: this images and text are auto-generated placeholders
and also the repos directory structure is gonna need a change
since this is typescript
I chose typescript because yesterday I did 4 rewrites in JS for the new api and all of them had some sort of issue
but with typescript I got no issue and it worked on the first try

**Reply by ProximalElk6186** [reply]
Wow, already looks pretty similar to the original!
About the change of repo's structure: if it needs to change, change it.

**Comment by 5d-code** [comment]
currently working on xmb items that can change background and show game covers etc

**Reply by ProximalElk6186** [reply]
Oh, cool!

**Reply by 5d-code** [reply]
after we are done with that we can add the final boss of xmb: submenus

**Reply by ProximalElk6186** [reply]
Yeah. They're supposed to be hidden all the time but not when you are accessing them.

**Comment by 5d-code** [comment]
also whats ur timezone it will be useful for communication mine is athens

**Reply by ProximalElk6186** [reply]
Mine is Tallinn.

**Reply by 5d-code** [reply]
actually we have the exact same time I think

**Reply by ProximalElk6186** [reply]
True. And our countries both do the summer time switch

**Comment by 5d-code** [comment]
hey i was guessing lets keep this on the web browser, so devices that can run a browser but cant run other programs (rokus, android tvs and ironically the ps4) and make custom server software that would serve the UI but would have some special routes in order to stream media, manage files, etc. but doing things like this over the network is kind of dangerous, we need to implement some sort of auth.

**Reply by ProximalElk6186** [reply]
So, you say that the UI should run on a dedicated server?

**Reply by 5d-code** [reply]
PS: and why i haven't commited something yet; I am trying to make a mockup of the PS3 menus but the icon rendering is off. due to this codebase being based on something i cobbled together in 30 minutes. I tried to add constants to reduce the amount of magic values, but I kept some because I randomly added them and they seemed to fix the graphical bugs. anyways how about you work on the server and I can work on the UI?

**Reply by 5d-code** [reply]
yes

**Reply by 5d-code** [reply]
it would open up the possibility of restricted devices that just have a web browser to run this. also for launching games: we can force the user to run the server on localhost, and connect to it through his web browser and the server whould open up the game windows, since you cant open up programs installed on server to client

**Reply by ProximalElk6186** [reply]
> PS: and why i haven't commited something yet; I am trying to make a mockup of the PS3 menus but the icon rendering is off. due to this codebase being based on something i cobbled together in 30 minutes. I tried to add constants to reduce the amount of magic values, but I kept some because I randomly added them and they seemed to fix the graphical bugs. anyways how about you work on the server and I can work on the UI?

Ok.

**Reply by ProximalElk6186** [reply]
> it would open up the possibility of restricted devices that just have a web browser to run this. also for launching games: we can force the user to run the server on localhost, and connect to it through his web browser and the server whould open up the game windows, since you cant open up programs installed on server to client

Sounds like it may actually work

**Reply by ProximalElk6186** [reply]
But on devices like Android TVs and the PS4 you cannot get the localhost

**Reply by 5d-code** [reply]
actually, lets just make the client electron/native. it would be way much of a hassle to make it run on a server now that i think of it. but we could allow it to connect to a plex server

**Reply by ProximalElk6186** [reply]
Ok, considering that the original goal was to make an XMB **LINUX** Media Center, and not cross-platform

**Reply by 5d-code** [reply]
yeah true

**Comment by ProximalElk6186** [comment]
Hi!

I've got a prototype of the options menu in p5.js

The code is dirty because I wanted to add an animation, but no.
```
let areOptionsShownOnScreen = false;
let optRectInitPos = 400;
let optTextInitPos = 400;
const options = ["Start", "Properties", "4732890", "875273485034", "", "423403219", "4238749302074"]; // Options stored in a constant array

function setup() 
{
  createCanvas(400, 400);
}

function draw() 
{
  background(220); // Redraw background to clear previous frames

  if (areOptionsShownOnScreen) 
  {
    // Play animation
    
    let optCurrPos = optRectInitPos;
    let optTextCurrPos = optTextInitPos;
    
        optCurrPos = 200;
    optTextCurrPos = 250;
    
    // Draw options menu if toggled on
    fill(180);
    rect(optCurrPos, 0, 200, 400);
    
    // Draw each option from the array
    fill(0);
    textSize(18);
    for (let i = 0; i < options.length; i++) 
    {
      text(options[i], optTextCurrPos, 50 + i * 50);
    }
    

    


    
  }
}

function keyPressed() 
{
  if (keyCode === ENTER) 
  {
    areOptionsShownOnScreen = !areOptionsShownOnScreen;
  }
}
```

__Theoretically__ it should work
